### PR TITLE
adding best buy review scraping functionality

### PR DESCRIPTION
**Pull request by kprazuch**
## Link to Ticket
https://trello.com/c/MyKC5E1x/107-scrape-reviews-from-best-buy

## Brief Description of Changes

- Added Best Buy reviews to endpoint /api/bestbuy-product
- Added additional package to requirements.txt
- Added unit tests for endpoint
- Updated README with additional step to install NLTK dependencies

NOTE: Top positive and negative reviews are not provided by this API so I'm going to make another branch/PR for this functionality specifically.

## Checklist
- [x] I do not commit the .env file.
- [x] I added my changes to the appropriate frontend or backend folders.
- [x] All aspects of the ticket have been entirely addressed and completed.
- [x] Thorough automated tests have been added.
- [x] Functionality has been validated through manual tests.
- [ ] I have received approval from a peer before merging.

### CODE REVIEW

File: backend/src/scraper/bestbuy_scraper.py [hunk 1]
```diff
@@ -45,12 +48,112 @@ def _product_specs(product_id: str) -> Dict:
...
+            l2 =[] 
+            
+            name_1_set = {ps.stem(w) for w in name_1 if not w in sw}  
+            name_2_set = {ps.stem(w) for w in name_2 if not w in sw} 
+            rvector = name_1_set.union(name_2_set)
+            for w in rvector: 
+                if w in name_1_set: l1.append(1)
+                else: l1.append(0) 
+                if w in name_2_set: l2.append(1) 
+                else: l2.append(0)
```
Comment (tzada8) [inline backend/src/scraper/bestbuy_scraper.py:89]: nit (don't feel like you have to make this change, but just a formatting tip):

```suggestion
                l1.append(1) if w in name_1_set else l1.append(0) 
                l2.append(1) if w in name_2_set else l2.append(0) 
```

**Review by tzada8 — COMMENTED** [review]
Can you explain the high-level flow to me? Like the idea behind the stemmer and cosine similarity?

Also, can you remind me what the API limits are for Best Buy? It looks like we'll be hitting their API 12 times for a single product, so just want to confirm if this might cause any problems

File: README.md [hunk 2]
```diff
@@ -25,8 +25,9 @@ Setup Backend:
 3. Create local virtual environment: `python3 -m venv .venv` (or use `python`)
 4. Activate virtual environment: `source .venv/bin/activate`
 5. Install dependencies: `pip install -r requirements.txt`
-6. Request environment variables from team: `.env` file
-7. Deactivate local virtual environment (when needed): `deactivate`
+6. Install NLTK dependencies: `python3 -m nltk.downloader punkt` and `python3 -m nltk.downloader stopwords` (if you are running into an error, try running `pip install certifi /Applications/Python\ {PYTHON_VERSION}/Install\ Certificates.command` first and then re-installing the NLTK dependencies)
```
[the remark this answers is not in the document]
Comment (kprazuch) [reply README.md:28]: Yeah, it's different. It doesn't seem possible to download through the requirements file.

**Comment by kprazuch** [comment]
> Can you explain the high-level flow to me? Like the idea behind the stemmer and cosine similarity?

Basically, this is a different Best Buy API than the one used for product specifications. This one is developed by Best Buy Canada whereas the other is by Best Buy US. Also, it's important to note that the SKUs used by Best Buy US and Best Buy Canada are different which is why I can't pass the SKU from specs to reviews. So instead, we first search for the product by UPC (which is consistent across countries) and then use the SKU that is returned by that endpoint to scrape the reviews. The reason I implemented cosine similarity is that I wanted to ensure that the product being returned by the search step matches the product that the user is actually looking at. Since the SKUs are different between the spec and review APIs, I decided that the product name was the easiest to compare. But, SURPRISE! The names aren't consistent between countries either. For example, 'Sony - ZV-1F Vlog Camera for Content Creators and Vloggers - Black' is the American product name and 'Sony ZV-1F Content Creator Vlogger 20.1MP Digital Camera - Black' is the Canadian product name. In this case, the names seem pretty similar but when you calculate cosine similarity on them, words that are essentially the same get counted as different (ex. creators vs. creator; vloggers vs. vlogger). Hence, the stemmer.

> Also, can you remind me what the API limits are for Best Buy? It looks like we'll be hitting their API 12 times for a single product, so just want to confirm if this might cause any problems

This API is different than the specifications one and much more under the radar (no longer updated and hard-to-find documentation). But, I haven't heard/read about a limit for this API so fingers crossed there isn't one. The other Best Buy API has a limit of 5 queries per second or 50,000 queries per day.

**Comment by tzada8** [comment]
Awesome, that makes sense. Thanks for this detailed response!

And 50k queries/day is more than enough haha! That's great

**Review by tzada8 — APPROVED** [review]

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (tzada8) [reply README.md:28]: What am I doing wrong here?

Screenshot 2024-01-29 at 6 34 17 PM

[the remark this answers is not in the document]
Comment (kprazuch) [reply README.md:28]: Hmmm try:
```
>> pip install certifi                                                
/Applications/Python\ {PYTHON_VERSION}/Install\ Certificates.command
```
It's one command split across two lines. Also make sure that the Python version is the exact same as listed in your File Explorer?

[the remark this answers is not in the document]
Comment (tzada8) [reply README.md:28]: Got it downloaded, thanks!

Also, just curious, are you manually adding the lines to the `requirements.txt` file?

[the remark this answers is not in the document]
Comment (kprazuch) [reply README.md:28]: Um yeah

**Pull request merged**